home-environment: allow adding to $PATH

afontaine opened the pull request:
### Description

Sometimes I'll have some things installed to my dot files that provide
a collection of executables (doom-emacs, org-capture), or need a
`$HOME/bin` or similar to keep a handful of quick scripts and wanted to
be able to do it decrlaratively.

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.

- [X] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

afontaine commented:
Hmm looks like the build failed because fetching `etex` failed, which is very unrelated. Can we try restarting the build to see if the `etex` servers are awake?

turion commented:
Does that work for you locally?

I tried `home.sessionVariables.PATH = "$PATH:/foo/bar";`, which should have given me an equivalent result (and it does produce the same `.nix-profile/etc/profile.d/hm-session-vars.sh` as yours), but `PATH` is not modified by that.

turion commented:
> Hmm looks like the build failed because fetching `etex` failed, which is very unrelated. Can we try restarting the build to see if the `etex` servers are awake?

Either @rycee can restart the build on Travis or you can make some change and force-push, that might trigger the build as well.

turion commented:
> I tried `home.sessionVariables.PATH = "$PATH:/foo/bar";`, which should have given me an equivalent result (and it does produce the same `.nix-profile/etc/profile.d/hm-session-vars.sh` as yours), but `PATH` is not modified by that.

Hmm, I think I'm wrong and that does indeed work. I just needed to restart my system.

turion commented on modules/home-environment.nix:
```diff
@@ -206,6 +206,12 @@ in
+    home.sessionPath = mkOption {
+      type = types.envVar;
+      description = "Extra directories to add to the $PATH";
+      default = "";
```
Can you give an example value with multiple paths?

afontaine replied:
updated :+1:

turion commented on modules/home-environment.nix:
```diff
@@ -374,7 +380,11 @@ in
             ${config.lib.shell.exportAll cfg.sessionVariables}
-          '' + cfg.sessionVariablesExtra;
+          ''
+          + (if cfg.sessionPath != "" then ''
```
Is it possible to pass a list of paths as well, and you concatenate the list with `:`?

afontaine replied:
I'm not sure of Nix's union-type capabilities, but the [envVar](https://github.com/NixOS/nixpkgs/blob/master/lib/types.nix#L221) type is simply a string. The only benefit is that the paths are concatenated together with a `:` on merging of the values. If you wish to add many paths in a single call, your best bet is to do something like

```nix
home.sessionPath = "foo/bar:foo/baz";
```

A `listOf` approach is certainly doable, as those are merged together as well, but I'm not sure which is preferential. We'd have to do the concatenation manually instead of relying on the type to do it for us.

berbiche replied:
`lib.strings` already provides `makeBinPath` to construct a `$PATH`

afontaine replied:
`makeBinPath` appends `/bin` to the given values, which isn't necessarily what a user might want.

I've changed it to a list that uses `concatStringsSep` to allow using a `listOf str` instead.

teto commented:
For reference, Nixos has the weird limited option:
```
environment.homeBinInPath
     Include ~/bin/ in $PATH.
```
what if you want to prepend to PATH instead ?  can't you just modify `home.sessionVariables` instead ? I don't believe we need a specific option for that.

turion commented:
@teto 

> what if you want to prepend to PATH instead ? 

Does that have a different effect:

> can't you just modify `home.sessionVariables` instead ? I don't believe we need a specific option for that.

See my previous comment: https://github.com/rycee/home-manager/pull/1292#issuecomment-638785530 It doesn't work straightforwardly.

teto commented:
> Hmm, I think I'm wrong and that does indeed work. I just needed to restart my system.

Seems like it worked ?! this behaves no differently: if you change sessions variables, you have to relog, otherwise, you could change it into your shell variables.

turion commented:
> > Hmm, I think I'm wrong and that does indeed work. I just needed to restart my system.
> 
> Seems like it worked ?!

You're right, I forgot!

afontaine commented:
@turion sorry for letting this stall for so long, but I've updated the documentation a bit more and fixed the default example.

As for _why_ this and not just `home.sessionVariables.PATH`, this approach allows users to break declarations apart and use them across files/modules/etc if you aren't about having one long file for home-manager.

I have `~/.config/emacs/bin` in my path for Doom Emacs (as it is not yet very nix-able), and [`.git/safe/../../bin`](https://thoughtbot.com/blog/git-safe), primarily for working with non-nix'd projects. These belong in the emacs set-up file I have and the git setup file, respectively.

The prepend/append issue is interesting, but I'm not sure what the best approach is to this beyond complicating this even further with something like `{path = "some/path"; prepend = true;}`

turion commented on modules/home-environment.nix:
```diff
@@ -206,6 +206,13 @@ in
+    home.sessionPath = mkOption {
+      type = with types; listOf str;
+      description = "Extra directories to add to the $PATH";
+      example = [ ".git/safe/../../bin" "some/other/path" ];
```
These aren't relative paths, right? Or relative to what are they? I find the examples a bit hard to grasp. Wouldn't a typical application of this option have `~/.local/bin` or so?

afontaine replied:
They are, in fact, relative paths! They can be whatever, after all. I should maybe change this to just be what I am currently using it for:

```suggestion
      example = [ ".git/safe/../../bin" "$(xdg.configHome}/emacs/bin" ];
```

turion replied:
Relative to _what_? To the current source file? To `hm-session-vars.sh`? Can you add that to the documentation? I like the second example, but I wouldn't know what the first example does.

afontaine replied:
if a relative path is in your `$PATH`, it is relative to your current working directory.

let's say I am working on a node app. if my `PATH` is `node_modules/.bin`, then when I am `cd`'d into the repo's directory, `REPO/node_modules/.bin` is available without using `npx` or similar. Instead of calling `npx eslint` or whatever, you just call `eslint`.

These values are just plopped into the PATH as-is, so there isn't really much to document other than that.

Maybe the description line can read:

`Extra directories appended to $PATH. Added as-is.` Maybe my example just doesn't make sense for most people's use case?

turion replied:
```suggestion
      example = [
        ".git/safe/../../bin" # Relative paths are expanded in the current working directory
        "$(xdg.configHome}/emacs/bin" # Store paths are expanded to absolute paths
        "~/.local/bin" # Bash and other modern shells expand ~ to your home directory
      ];
```
Often a well-documented example suffices. How about something like this?

> if a relative path is in your $PATH, it is relative to your current working directory.

Oh, I didn't know!

afontaine replied:
Ah yeah _much_ better, thank you :bow:

and yeah, it is a good trick sometimes :grin:

turion commented:
I think this is a good option to add.

@rycee probably your input is needed here :)

teto commented:
one can already adjust the PATH via `export PATH=/my/bin:${PATH}` so this looks overengineered ?

afontaine commented:
> one can already adjust the PATH via `export PATH=/my/bin:${PATH}` so this looks overengineered ?

How? Where? Through `sessionVariables`? in `myShell.initExtra`?

If in either of those options, how do we append to it across multiple `nix` files? do I need to keep track of all the path adjustments I want to make and keep them in one location? That feels strange.

Admittedly, it does feel slightly overkill, but at the same time, the current solutions feel... lacking. I'm not really sure what the best solution is though.

turion commented:
I don't think this is overkill or overengineered. It abstracts adding directories to the PATH from the actual mechanism. It's good that one can configure the PATH variable without having to read the home-manager source code in order to understand where and when and how `hm-session-vars.sh` is called. And it probably does no harm because it has a unit test.

rycee commented:
Perhaps it would suffice to fix the type of `sessionVariables` to be a `attrsOf` instead of `attrs` with a list value being concatenated by `:`? Something like
```diff
diff --git a/modules/home-environment.nix b/modules/home-environment.nix
index 209df4ff..7691e6a1 100644
--- a/modules/home-environment.nix
+++ b/modules/home-environment.nix
@@ -228,7 +228,11 @@ in
 
     home.sessionVariables = mkOption {
       default = {};
-      type = types.attrs;
+      type = with types;
+        let
+          primitive = either bool (either int str);
+          var = either primitive (listOf str);
+        in attrsOf var;
       example = { EDITOR = "emacs"; GS_OPTIONS = "-sPAPERSIZE=a4"; };
       description = ''
         Environment variables to always set at login.
diff --git a/modules/lib/shell.nix b/modules/lib/shell.nix
index 5e5743f5..7b57596d 100644
--- a/modules/lib/shell.nix
+++ b/modules/lib/shell.nix
@@ -2,7 +2,10 @@
 
 rec {
   # Produces a Bourne shell like variable export statement.
-  export = n: v: ''export ${n}="${toString v}"'';
+  export = let
+    toShell = v:
+      if lib.isList v then lib.concatStringsSep ":" v else toString v;
+  in n: v: ''export ${n}="${toShell v}"'';
 
   # Given an attribute set containing shell variable names and their
   # assignment, this function produces a string containing an export
```

Then the configuration
```
home.sessionVariables = lib.mkMerge [
  (lib.mkAfter {
    PATH = ["$PATH"];
  })
  {
    PATH = ["$HOME/path1"];
  }
  {
    PATH = ["$HOME/path2"];
  }
];
```
would place
```sh
export PATH="$HOME/path1:$HOME/path2:$PATH"
```
in `hm-session-vars.sh`.

It wouldn't correctly handle an empty `PATH` variable, though.

berbiche commented:
@rycee Quick question: why would you use `either` when `listOf` is available? Are the documentation/error strings clearer?

rycee commented:
@berbiche I assume you mean `types.oneOf`. It's mainly out of habit. Probably better to use `oneOf` since it is a more stylish and the generated code should be identical to using `either` explicitly.

afontaine commented:
> Perhaps it would suffice to fix the type of `sessionVariables` to be a
> `attrsOf` instead of `attrs` with a list value being concatenated by `:`?

This was my initial thought as well, @rycee, but I am relatively new to writing nix so
I'm not too familiar with how to go about it. I initially just started by trying
to edit `toShell` to work with lists as well.

> It wouldn't correctly handle an empty `PATH` variable, though.

This was my big blocker though. `PATH` is such a frustratingly special case, and
I couldn't find an elegant way to inject a `:$PATH` into the variable building
logic, so it felt easier to split it out into a separate option. It is also
required here unfortunately :disappointed:

turion commented:
> Perhaps it would suffice to fix the type of `sessionVariables` to be a `attrsOf` instead of `attrs` with a list value being concatenated by `:`?

But there are lots of session variables that aren't paths.

marsam commented on modules/home-environment.nix:
```diff
@@ -206,6 +206,17 @@ in
 
+    home.sessionPath = mkOption {
+      type = with types; listOf str;
+      description = "Extra directories to add to the $PATH";
```
```suggestion
      description = "Extra directories to add to the <literal>PATH</literal> environment variable.";
```

marsam commented on modules/home-environment.nix:
```diff
@@ -374,7 +385,11 @@ in
 
             ${config.lib.shell.exportAll cfg.sessionVariables}
-          '' + cfg.sessionVariablesExtra;
+          ''
+          + (if cfg.sessionPath != "" then ''
+            export PATH="$PATH:${builtins.concatStringsSep ":" cfg.sessionPath}"
+          '' else "")
+          + cfg.sessionVariablesExtra;
```
I think we can use
```suggestion
          '' + lib.optionalString (cfg.sessionPath != []) ''
            export PATH="$PATH''${PATH:+:}${builtins.concatStringsSep ":" cfg.sessionPath}"
          '' + cfg.sessionVariablesExtra;
```

afontaine replied:
Yes that looks much better :+1:

afontaine commented on tests/modules/home-environment/session-path-expected.txt:
```diff
@@ -0,0 +1,8 @@
+export XDG_CACHE_HOME="/home/hm-user/.cache"
+export XDG_CONFIG_HOME="/home/hm-user/.config"
+export XDG_DATA_HOME="/home/hm-user/.local/share"
+export PATH="$PATH${PATH:+:}:bar:baz:foo"
```
```suggestion
export PATH="$PATH${PATH:+:}bar:baz:foo"
```

afontaine commented:
@rycee is there anything I can change here to move this along, aside from merge conflicts?

rycee commented on modules/home-environment.nix:
```diff
@@ -263,6 +263,17 @@ in
+      description = "Extra directories to add to the <literal>PATH</literal>";
+      example = [
+        ".git/safe/../../bin" # Relative paths are expanded in the current working directory
+        "$(xdg.configHome}/emacs/bin" # Store paths are expanded to absolute paths
```
```suggestion
        "\${config.xdg.configHome}/emacs/bin"
```

rycee commented on modules/misc/news.nix:
```diff
@@ -1527,7 +1527,6 @@ in
           A new module is available: 'programs.zoxide'
         '';
       }
-
```
Removes newline

rycee commented:
I made some minor changes and merged to master. Thanks for the contribution!

afontaine commented:
awesome, thanks @rycee
